Richer Contacts API

grabbou opened the pull request:
Fixes #18. Implements richer contacts API.

Here's the current preview of what's outputted: https://gist.github.com/grabbou/5272743f4b0c08cb3279f28d4996b21f

Tested with missing street/city and other properties on iOS.

grabbou committed:
Initial commit

grabbou committed:
Fix crashes

grabbou commented on ios/Exponent/Versioned/Modules/Api/EXContacts.m:
```diff
@@ -70,31 +68,60 @@ - (void)_getContactsWithPermissionGrantedAsync: (NSArray *)fields addressBook:(A
-    NSString *lastName = (__bridge_transfer NSString *)(ABRecordCopyValue(person, kABPersonLastNameProperty));
-    NSString *name = [self _assembleDisplayNameFromFirstName:firstName lastName:lastName];
-    if (name) {
-      contact[@"name"] = name;
```
Not sure if we need to check for value, AFAIK passing `null` should work via bridge.

grabbou committed:
Remove legacy as per brent suggestion

grabbou committed:
Protect against nil release

grabbou committed:
Move around

grabbou commented on ios/Exponent/Versioned/Modules/Api/EXContacts.m:
```diff
@@ -53,48 +53,77 @@ - (void)dealloc
+          
+          CFStringRef emailLabelRef = ABMultiValueCopyLabelAtIndex(emails, index);
+          NSString *emailLabel = emailLabelRef ?
+            (__bridge_transfer NSString *)(ABAddressBookCopyLocalizedLabel(emailLabelRef)) :
```
I've been always indenting on tenary, like here. Xcode keeps aligning that line with `NSString` on 114, is that against our conventions?

satya164 commented:
Instead of a `phoneNumber` field and a `phoneNumbers` field, does it make more sense to to add a field called `default: true` to the default phone number?

satya164 commented:
> I am not really sure why we need to pass fields array to the native side

Last time I tried on Android, fetching contacts was very slow. Maybe because of that?

grabbou commented:
@satya164 I guess the default number is first item from an array? Anyways, I removed `phoneNumber` since we have an array. Not sure if returning first item from it separately makes sense tho

satya164 commented:
@grabbou What if there is no default though?

grabbou commented:
What you mean? AFAIK there's no explicit default, at least on iOS. That's why Exponent used to return `phoneNumber` as a first item from an array (at least on iOS). I removed that and replaced in favour of an array - which seems more reasonable. Initially I kept the `phoneNumber` as is to provide non-breaking change, but Brent said it's irrelevant and easy to upgrade, so lets' do it

satya164 commented:
@grabbou on Android you can set a number as default

grabbou commented:
Ok, you are right, then I should either use flag or `phoneNumber` (I guess latter makes more sense). I'll check android and play around with naming. Thanks!

grabbou committed:
Bring back if

grabbou committed:
wat

grabbou committed:
Clean up diff

grabbou committed:
On iOS, make first element default

satya164 committed:
Add ability to get addresses, emails and phone numbers from contact

satya164 committed:
Merge branch 'feat/rich-contacts' of github.com:satya164/exponent into feat/rich-contacts

grabbou committed:
Align API

satya164 committed:
Add company and job title

satya164 committed:
Fix typos

satya164 committed:
Fix querying company details

grabbou committed:
Fixes

grabbou committed:
Merge branch 'feat/rich-contacts' of github.com:satya164/exponent into feat/rich-contacts

grabbou committed:
Remove hard to maintain comment

terribleben commented:
Thanks guys!

brentvatne commented:
@grabbou @satya164 - This appears to be missing image data [ABPersonCopyImageData](http://stackoverflow.com/questions/2085959/get-image-of-a-person-from-iphone-address-book) -- was that intentional? Not sure if a similar thing is possible on Android. Also, not sure if Android can return `firstName`, `middleName`, and `lastName` but that would be good for consistency.

grabbou commented:
We delayed implementation of an image to next iteration (aka follow-up PR) to get this out faster (there was a deadline of a day I guess before the release).

satya164 commented:
@brentvatne 

> This appears to be missing image data ABPersonCopyImageData

What's structure of the data JS expects? On Android we can get URI of the person's avatar

> Also, not sure if Android can return firstName, middleName, and lastName but that would be good for consistency

Can't, android only has display name (also that's what we enter when saving contacts). Do people have to enter these 3 separate fields on iOS when adding a new contact?

brentvatne commented:
@satya164

> What's structure of the data JS expects? On Android we can get URI of the person's avatar

I think the URI is fine!

> Can't, android only has display name (also that's what we enter when saving contacts). Do people have to enter these 3 separate fields on iOS when adding a new contact?

Ah ok! I'm not sure how to add the middle name on iOS but I guess it must be possible somehow. First name and last name are separate though.
